kvserver: fail gracefully in TestLeaseTransferRejectedIfTargetNeedsSnapshot

tbg opened the pull request:
We saw this test hang in CI. What likely happened (according to the stacks) is
that a lease transfer that was supposed to be caught by an interceptor never
showed up in the interceptor. The most likely explanation is that it errored
out before it got to evaluation. It then signaled a channel the test was only
prepared to check later, so the test hung (waiting for a channel that was now
never to be touched).

This test is hard to maintain. It would be great (though, for now, out of reach)
to write tests like it in a deterministic framework[^1]

[^1]: see https://github.com/cockroachdb/cockroach/issues/105177.

For now, fix the test so that when the (so far unknown) error rears its
head again, it will fail properly, so we get to see the error and can
take another pass at fixing the test (separately). Stressing
this commit[^2], we get:

> transferErrC unexpectedly signaled: /Table/Max: transfer lease unexpected
> error: refusing to transfer lease to (n3,s3):3 because target may need a Raft
> snapshot: replica in StateProbe

This makes sense. The test wants to exercise the below-raft mechanism, but
the above-raft mechanism also exists and while we didn't want to interact
with it, we sometimes do[^1]

The second commit introduces a testing knob that disables the above-raft
mechanism selectively. I've stressed the test for 15 minutes without issues
after this change.

[^1]: somewhat related to https://github.com/cockroachdb/cockroach/issues/107524
[^2]: `./dev test --filter TestLeaseTransferRejectedIfTargetNeedsSnapshot --stress ./pkg/kv/kvserver/` on gceworker, 285s

Fixes #106383.

Epic: None
Release note: None

tbg committed:
kvserver: fail gracefully in TestLeaseTransferRejectedIfTargetNeedsSnapshot

We saw this test hang in CI. What likely happened (according to the stacks) is
that a lease transfer that was supposed to be caught by an interceptor never
showed up in the interceptor. The most likely explanation is that it errored
out before it got to evaluation. It then signaled a channel the test was only
prepared to check later, so the test hung (waiting for a channel that was now
never to be touched).

This test is hard to maintain. It would be great (though, for now, out of reach)
to write tests like it in a deterministic framework[^1]

[^1]: see https://github.com/cockroachdb/cockroach/issues/105177.

For now, fix the test so that when the (so far unknown) error rears its
head again, it will fail properly, so we get to see the error and can
take another pass at fixing the test (separately). Stressing
this commit[^2], we get:

> transferErrC unexpectedly signaled: /Table/Max: transfer lease unexpected
> error: refusing to transfer lease to (n3,s3):3 because target may need a Raft
> snapshot: replica in StateProbe

This makes sense. The test wants to exercise the below-raft mechanism, but
the above-raft mechanism also exists and while we didn't want to interact
with it, we sometimes do[^1]

[^1]: somewhat related to https://github.com/cockroachdb/cockroach/issues/107524
[^2]: `./dev test --filter TestLeaseTransferRejectedIfTargetNeedsSnapshot --stress ./pkg/kv/kvserver/` on gceworker, 285s

Touches #106383.

Epic: None
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/107526)
<!-- Reviewable:end -->

tbg committed:
kvserver: deflake TestLeaseTransferRejectedIfTargetNeedsSnapshot

See previous commit. We sometimes hit the above-raft check when we wanted
to hit only the below-raft one.

This commit fixes this by selectively disabling the above-raft check in
this test.

Note that not all tests using lease transfers are susceptible to this problem
in the way that this test is. This is because this test also lowers the
`LeaseTransferRejectedRetryLoopCount`[^1] because it is intentionally
manufacturing failed lease transfers and doesn't want to sit out the retry
loop. It is that time-saving optimization that also allows the spurious error
to bubble up.

[^1]: https://github.com/cockroachdb/cockroach/blob/66c9f93ae86bddd7ba4c5f6a6b8b6cb700ca23ce/pkg/kv/kvserver/testing_knobs.go#L375

Epic: none
Release note: None

erikgrinaker reviewed (approved):
The new testing knob isn't necessary, you can use `AdminTransferLeaseBypassingSafetyChecks`:

https://github.com/cockroachdb/cockroach/blob/adfa97f100ab02fec4a9211bd3a9acdd65755f56/pkg/kv/db.go#L666-L675

tbg commented:
> The new testing knob isn't necessary, you can use AdminTransferLeaseBypassingSafetyChecks:

That disables both checks (above and below raft), but I need the below-raft to stay active.

TFTR!

bors r=erikgrinaker
